iree/base/loop_inline.c

```diff
@@ -0,0 +1,514 @@
+      iree_wait_source_wait_one(params.wait_source, timeout);
+
+  // Callback after wait, whether it succeeded or failed.
+  iree_status_t status =
+      params.callback.fn(params.callback.user_data, loop, wait_status);
+  if (!iree_status_is_ok(status)) {
+    iree_loop_inline_emit_error(loop, status);
+  }
+
+  IREE_TRACE_ZONE_END(z0);
+}
+
+// IREE_LOOP_COMMAND_WAIT_ANY
+static void iree_loop_inline_run_wait_any(
+    iree_loop_t loop, iree_loop_wait_multi_params_t params) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_timeout_t timeout = iree_make_deadline(params.deadline_ns);
+
+  // Do a scan down the wait sources to see if any are already set - if so we
+  // can bail early. Otherwise we need to wait on any one.
+  // iree_wait_any is a much more efficient (and fair) way but this keeps the
+  // code working on bare-metal.
+  iree_status_t wait_status = iree_status_from_code(IREE_STATUS_DEFERRED);
+  for (iree_host_size_t i = 0; i < params.count; ++i) {
+    iree_status_code_t wait_status_code = IREE_STATUS_OK;
+    iree_status_t query_status =
+        iree_wait_source_query(params.wait_sources[i], &wait_status_code);
+    if (iree_status_is_ok(query_status)) {
+      if (wait_status_code == IREE_STATUS_OK) {
+        // Signaled - can bail early.
+        break;
+      } else if (wait_status_code == IREE_STATUS_DEFERRED) {
+        // Not signaled yet - keep scanning.
+        continue;
+      } else {
+        // Wait failed - can bail early.
+        wait_status = iree_status_from_code(wait_status_code);
+        break;
+      }
+    } else {
+      // Failed to perform the query, which we treat the same as a wait error.
+      wait_status = query_status;
+      break;
+    }
+  }
+  if (iree_status_is_deferred(wait_status)) {
+    // No queries resolved/failed - commit any real wait.
+    // We choose the first one to be (somewhat) deterministic but really it
+    // should be randomized... or if the user cares they should use a real loop.
+    wait_status = iree_wait_source_wait_one(params.wait_sources[0], timeout);
+  }
+
+  // Callback after wait, whether it succeeded or failed.
+  iree_status_t status =
+      params.callback.fn(params.callback.user_data, loop, wait_status);
+  if (!iree_status_is_ok(status)) {
+    iree_loop_inline_emit_error(loop, status);
+  }
+
+  IREE_TRACE_ZONE_END(z0);
+}
+
+// IREE_LOOP_COMMAND_WAIT_ALL
+static void iree_loop_inline_run_wait_all(
+    iree_loop_t loop, iree_loop_wait_multi_params_t params) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_timeout_t timeout = iree_make_deadline(params.deadline_ns);
+
+  // Run down the list waiting on each source.
+  // iree_wait_all is a much more efficient way but this keeps the code working
+  // on bare-metal.
+  iree_status_t wait_status = iree_ok_status();
+  for (iree_host_size_t i = 0; i < params.count; ++i) {
+    wait_status = iree_wait_source_wait_one(params.wait_sources[i], timeout);
+    if (!iree_status_is_ok(wait_status)) break;
+  }
```

Comment:
Could `iree_wait_all` on bare-metal just be implemented in this way? Then this code could take advantage of efficient implementations when available.

Reply:
the wait set stuff is currently too fat for that (requires KB of memory) - but my hope is to rework it so that it takes wait sources instead of wait handles and then we can have iree_wait_source_wait_all/any/etc that's much nicer